test/e2e/apps/deployment.go

```diff
@@ -1705,3 +1709,146 @@ func testDeploymentSubresources(ctx context.Context, f *framework.Framework) {
+			oldLabel, hasOld := pod.Labels[appsv1.DefaultDeploymentUniqueLabelKey]
+			newLabel, hasNew := pod.Labels[appsv1.DeploymentUniqueLabelKey]
+			if (hasNew != hasNewLabel) || (hasOld != hasOldLabel) {
+				return fmt.Errorf("pod doesn't have a hash label")
+			}
+			if (hasNew && hasOld) && (oldLabel != newLabel) {
+				return fmt.Errorf("pod has inconsistent hash label")
+			}
+		}
+		return nil
+	}
+
+	// create rs with old label
+	size := int32(1)
+	zero := int64(0)
+
+	rs := &appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-replicaset",
+			Namespace: ns,
+			Labels: map[string]string{
+				"test-name":                            "pod-template-hash-compatible",
+				appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+			},
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			Replicas: &size,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"test-name":                            "pod-template-hash-compatible",
+					appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"test-name":                            "pod-template-hash-compatible",
+						appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+					},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+
+	_, err := c.AppsV1().ReplicaSets(ns).Create(ctx, rs, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating replica set %q in namespace %q", rs.Name, ns)
+
+	err = e2ereplicaset.WaitForReadyReplicaSet(ctx, c, ns, rs.Name)
+	framework.ExpectNoError(err)
+
+	// create the deployment to control the replicaset
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-deployment",
+			Namespace: ns,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &size,
+			Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"test-name": "pod-template-hash-compatible"}},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"test-name": "pod-template-hash-compatible"},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+	deployment, err := c.AppsV1().Deployments(ns).Create(ctx, d, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+
+	if err := checkPodTemplateHash(ctx, 1, true, false); err != nil {
+		framework.Failf("failed checking labels after create the deployment, %v", err)
+	}
+
+	// scale should be ok with old label
+	newSize := int32(2)
+	d.Spec.Replicas = &newSize
+	deployment, err = c.AppsV1().Deployments(ns).Update(ctx, d, metav1.UpdateOptions{})
```

Comment:
scale endpoint can be also used

Reply:
Sorry, can you explain this in detail? I don't quite understand what you mean. What `scale endpoint`.

Reply:
take a look at `c.AppsV1().Deployments(ns).UpdateScale`